src/main/java/com/gotocompany/depot/maxcompute/schema/SchemaDifferenceUtils.java

```diff
@@ -0,0 +1,88 @@
+        List<MaxComputeColumnDetail> changedMetadata = new ArrayList<>();
+
+        while (newMaxComputeColumnDetailIterator.hasNext()) {
+            Map.Entry<String, MaxComputeColumnDetail> entry = newMaxComputeColumnDetailIterator.next();
+            String columnName = entry.getKey();
+            MaxComputeColumnDetail oldMetadata = oldMaxComputeColumnDetail.get(columnName);
+            if (Objects.isNull(oldMetadata)) { //handle new column / struct field
+                changedMetadata.add(entry.getValue());
+                if (isStructType(entry.getValue().getTypeInfo()) || isStructArrayType(entry.getValue().getTypeInfo())) {
+                    skipStructFields(entry, newMaxComputeColumnDetailIterator);
+                }
+            }
+        }
+        return changedMetadata;
+    }
+
+    private static void skipStructFields(Map.Entry<String, MaxComputeColumnDetail> entry, Iterator<Map.Entry<String, MaxComputeColumnDetail>> newMaxComputeColumnDetailIterator) {
+        StructTypeInfo structTypeInfo = isStructType(entry.getValue().getTypeInfo()) ? (StructTypeInfo) entry.getValue().getTypeInfo()
+                : ((StructTypeInfo) ((ArrayTypeInfo) entry.getValue().getTypeInfo()).getElementTypeInfo());
+        for (int i = 0; i < structTypeInfo.getFieldCount(); i++) {
+            newMaxComputeColumnDetailIterator.next();
+        }
+    }
+
+    private static Map<String, MaxComputeColumnDetail> buildMaxComputeColumnDetailMap(TableSchema schema) {
+        Map<String, MaxComputeColumnDetail> maxComputeColumnDetailMap = new TreeMap<>();
+        schema.getColumns().forEach(column -> fieldMetadataHelper(column.getTypeInfo(), "", column.getName(), maxComputeColumnDetailMap, false));
+        return maxComputeColumnDetailMap;
+    }
+
+    private static void fieldMetadataHelper(TypeInfo typeInfo, String prefix, String name, Map<String, MaxComputeColumnDetail> result, boolean isArrayElement) {
+        if (isPrimitiveType(typeInfo) || isPrimitiveArrayType(typeInfo)) {
+            result.put(getPathName(prefix, name, isArrayElement), new MaxComputeColumnDetail(prefix, name, typeInfo, isArrayElement));
+        }
+        if (isStructType(typeInfo) || isStructArrayType(typeInfo)) {
+            StructTypeInfo structTypeInfo = isStructType(typeInfo) ? (StructTypeInfo) typeInfo : ((StructTypeInfo) ((ArrayTypeInfo) typeInfo).getElementTypeInfo());
+            result.put(getPathName(prefix, name, isArrayElement), new MaxComputeColumnDetail(prefix, name, typeInfo, isArrayElement));
+
+            for (int i = 0; i < structTypeInfo.getFieldCount(); i++) {
+                TypeInfo fieldType = structTypeInfo.getFieldTypeInfos().get(i);
+                String fieldName = structTypeInfo.getFieldNames().get(i);
+                fieldMetadataHelper(fieldType, getPathName(prefix, name, isArrayElement), fieldName, result, isStructArrayType(typeInfo));
+            }
+        }
```

Comment:
The recursive schema difference computation could lead to stack overflow with deeply nested structures. Also lacks proper handling of column reordering and type changes.

Reply:
As of now, this is only a workaround until Ali's team provide a table update API that accepts the whole TableSchema.